SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ClosureSpecialization.swift

```diff
@@ -559,6 +567,390 @@ private func markConvertedAndReabstractedClosuresAsUsed(rootClosure: Value, conv
+    if self.cloned.blocks.isEmpty {
+      self.cloned.appendNewBlock(self.context)
+    } else {
+      self.cloned.entryBlock
+    }
+  }
+
+  func cloneAndSpecializeFunctionBody(using callSite: CallSite) {
+    self.cloneEntryBlockArgsWithoutOrigClosures(usingOrigCalleeAt: callSite)
+
+    let (allSpecializedEntryBlockArgs, closureArgIndexToAllClonedReleasableClosures) = cloneAllClosures(at: callSite)
+
+    self.cloneFunctionBody(from: callSite.applyCallee, entryBlockArgs: allSpecializedEntryBlockArgs)
+
+    self.insertCleanupCodeForClonedReleasableClosures(
+      from: callSite, closureArgIndexToAllClonedReleasableClosures: closureArgIndexToAllClonedReleasableClosures)
+  }
+
+  private func cloneEntryBlockArgsWithoutOrigClosures(usingOrigCalleeAt callSite: CallSite) {
+    let originalEntryBlock = callSite.applyCallee.entryBlock
+    let clonedFunction = self.cloned
+    let clonedEntryBlock = self.entryBlock
+
+    originalEntryBlock.arguments
+      .enumerated()
+      .filter { index, _ in !callSite.hasClosureArg(at: index) }
+      .forEach { _, arg in
+        let clonedEntryBlockArgType = clonedFunction.loweredType(of: arg.type)
+        let clonedEntryBlockArg = clonedEntryBlock.addFunctionArgument(type: clonedEntryBlockArgType, self.context)
+        clonedEntryBlockArg.copyFlags(of: arg as! FunctionArgument)
+      }
+  }
+
+  /// Clones all closures, originally passed to the callee at the given callSite, into the specialized function.
+  ///
+  /// Returns the following -
+  /// - allSpecializedEntryBlockArgs: Complete list of entry block arguments for the specialized function. This includes
+  ///   the original arguments to the function (minus the closure arguments) and the arguments representing the values
+  ///   originally captured by the skipped closure arguments.
+  ///
+  /// - closureArgIndexToAllClonedReleasableClosures: Mapping from a closure's argument index at `callSite` to the list
+  ///   of corresponding releasable closures cloned into the specialized function. We have a "list" because we clone
+  ///   "closure chains", which consist of a "root" closure and its conversions/reabstractions. This map is used to
+  ///   generate cleanup code for the cloned closures in the specialized function.
+  private func cloneAllClosures(at callSite: CallSite) 
+    -> (allSpecializedEntryBlockArgs: [Value], 
+        closureArgIndexToAllClonedReleasableClosures: [Int: [SingleValueInstruction]]) 
+  {
+    func entryBlockArgsWithOrigClosuresSkipped() -> [Value] {
+      var clonedNonClosureEntryBlockArgs = self.entryBlock.arguments.makeIterator()
+
+      return callSite.applyCallee
+        .entryBlock
+        .arguments
+        .enumerated()
+        .reduce(into: []) { result, origArgTuple in
+          let (index, _) = origArgTuple
+          if !callSite.hasClosureArg(at: index) {
+            result.append(clonedNonClosureEntryBlockArgs.next()!)
+          } else {
+            result.append(PlaceholderValue())
```

Comment:
I think it's better to append `nil` in this case and make the result an array of optional values